Add Travis-CI tests for testing SW counter measures against HW level fault injection

matetothpal opened the pull request:
The tests investigate the behaviour of MCUBoot in case of instruction skip. The idea behind the test implementation is to create a set of scripts that execute the following:
- Run in MCUBoot in QEMU with a corrupted image.
- Attach a debugger to QEMU, and stop execution of MCUBoot.
- Modify the PC register by adding a predefined value to it
- Continue the execution
- Check whether the image booted successfully.
- As it would be unrealistic to test with skipping all the instructions, the test is limited to the instructions that are hardened in MCUboot. The FIH_* macros define symbols in the code at the beginning and the end of the hardening code, and the address of those symbols are extracted by the test scripts.

To Achieve this, the pull requests consists of the following additions to the CI directory:
- A Dockerfile, that describes an image that can compile MCUBoot with TF-M, and can run it in QEMU.
- Scripts that run on the Travis test slave to build the docker image and compile MCUBoot with TF-M, and corrupt the boot image. Note: the image is corrupted by changing the signature of the image.
- Scripts that execute the instruction skip tests based on the parameters passed into the test environment from the .travis.yml file.
- A python script that creates a short summary of the test results based on the yaml file emitted by the test scripts.

Below are some statistics based on test results:

Release build (-O3):
```
+------------------------------+-----------------+----------------------+------------------------+
|                              | MCUBoot         | Number of skip tests | Successful boots with  |
|                              | Image Size      | executed             | corrupted image        |
+==============================+=================+======================+========================+
| MCUBOOT_FIH_PROFILE_NONE     | FLASH:  25.1 kB |         480          |         35             |
|                              | RAM:    25.4 kB |                      |                        |
+------------------------------+-----------------+----------------------+------------------------+
| MCUBOOT_FIH_PROFILE_LOW      | FLASH:  25.5 kB |         770          |         19             |
|                              | RAM:    25.4 kB |                      |                        |
+------------------------------+-----------------+----------------------+------------------------+
| MCUBOOT_FIH_PROFILE_MEDIUM   | FLASH:  27.8 kB |        1145          |          9             |
|                              | RAM:    25.4 kB |                      |                        |
+------------------------------+-----------------+----------------------+------------------------+
```

MinSizeRel build (-Os):
```
+------------------------------+-----------------+----------------------+------------------------+
|                              | MCUBoot         | Number of skip tests | Successful boots with  |
|                              | Image Size      | executed             | corrupted image        |
+==============================+=================+======================+========================+
| MCUBOOT_FIH_PROFILE_NONE     | FLASH:  24.3 kB |          540         |           35           |
|                              | RAM:    25.4 kB |                      |                        |
+------------------------------+-----------------+----------------------+------------------------+
| MCUBOOT_FIH_PROFILE_LOW      | FLASH:  24.7 kB |          840         |           24           |
|                              | RAM:    25.4 kB |                      |                        |
+------------------------------+-----------------+----------------------+------------------------+
| MCUBOOT_FIH_PROFILE_MEDIUM   | FLASH:  27.0 kB |         1220         |           12           |
|                              | RAM:    25.4 kB |                      |                        |
+------------------------------+-----------------+----------------------+------------------------+
```

Note: MCUBOOT_FIH_PROFILE_MAX is not tested as it requires RNG, and the AN521 platform in TF-M has no support for it. However this profile adds the random execution delay to the code, so should not affect the instruction skip results too much.

The tests executed by Travis for this pull request can be looked at to see the tests in operation.

Note: the test cases defined in .travis.yml always return passed, if they were executed successfully.

utzig commented:
Please rebase.

matetothpal commented:
I rebased the change on master.

However I'm a bit confused, because in this page it seems that the FIH test didn't run. Although if I have a look at Travis set up for my fork , those tests run as well: https://travis-ci.com/github/matetothpal/mcuboot/builds/188328067

I seem to recall that for earlier versions of the pull request I saw the tests that were added in the pull request being run as well.

utzig commented:
Yes, I disabled Travis but I will re-enable for this particular test.

matetothpal commented:
The changes necessary to run with the FI Hardened MCUBoot have been merged to TF-M, so removed the extra patching step from the scripts

utzig commented:
@matetothpal Sorry I got involved in other things yesterday and ended up forgetting this, can you do me a favor and send another PR disabling all the simulator and Mynewt build tests to Travis? I'd rather remote them first and then run only this PR's commits (and signed-off-by checks) on Travis while under review and later on master once approved.

matetothpal commented:
@utzig No problem. I created the PR: https://github.com/JuulLabs-OSS/mcuboot/pull/836

utzig commented:
@matetothpal Please rebase!

matetothpal commented:
> @matetothpal Please rebase!

@utzig Done

utzig commented:
@matetothpal `ci/fih_test_docker/Dockerfile` is using Windows line termination, could you do some `dos2unix` thing and push again?

utzig commented on ci/fih-tests_install.sh:
```diff
@@ -0,0 +1,30 @@
+./build.sh
+popd
+
+
```
Please remove all those empty lines existing at the end of the files, including some of the other scripts and the Dockerfile.

matetothpal commented:
@utzig I fixed the line ending, and removed the linebreaks from the end of the files. Also rebased to the latest master

utzig commented on ci/fih_test_docker/damage_image.py:
```diff
@@ -0,0 +1,234 @@
+    #print (tlv_info.pretty_str())
+    if tlv_info.it_magic == IMAGE_TLV_PROT_INFO_MAGIC:
+        logging.debug("Protected TLV found at offset 0x{:X}".format(tlv_info_offset))
+        if (image_header.ih_protect_tlv_size != tlv_info.it_tlv_tot):
```
No need for `(` and `)`

utzig commented on ci/fih_test_docker/damage_image.py:
```diff
@@ -0,0 +1,234 @@
+        logging.debug("Protected TLV found at offset 0x{:X}".format(tlv_info_offset))
+        if (image_header.ih_protect_tlv_size != tlv_info.it_tlv_tot):
+            logging.error("Invalid prot TLV len ({:d} vs. {:d})".format(image_header.ih_protect_tlv_size, tlv_info.it_tlv_tot))
+            return
```
Should `raise` instead

utzig commented on ci/fih_test_docker/damage_image.py:
```diff
@@ -0,0 +1,234 @@
+        tlv_info = ImageTLVInfo.read_from_binary(in_file)
+
+    else:
+        if (image_header.ih_protect_tlv_size != 0):
```
No need for `(` and `)`

utzig commented on ci/fih_test_docker/damage_image.py:
```diff
@@ -0,0 +1,234 @@
+    else:
+        if (image_header.ih_protect_tlv_size != 0):
+            logging.error("No prot TLV was found.")
+            return
```
Raise

utzig commented on ci/fih_test_docker/damage_image.py:
```diff
@@ -0,0 +1,234 @@
+    logging.debug("Unprotected TLV found at offset 0x{:X}".format(tlv_info_offset))
+    if tlv_info.it_magic != IMAGE_TLV_INFO_MAGIC:
+        logging.error("Invalid magic in tlv info: 0x{:X} instead of 0x{:X}".format(tlv_info.it_magic, IMAGE_TLV_INFO_MAGIC))
+        return
```
Raise

utzig commented on ci/fih_test_docker/damage_image.py:
```diff
@@ -0,0 +1,234 @@
+        logging.error("Invalid magic in tlv info: 0x{:X} instead of 0x{:X}".format(tlv_info.it_magic, IMAGE_TLV_INFO_MAGIC))
+        return
+
+
```
One empty line too much

utzig commented on ci/fih_test_docker/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+        pyasn1==0.1.9 &&\
+    # Install toolchain
+    mkdir -p /root/bin/gcc-7-2018-q2-update &&\
+    curl --create-dirs -fsSLo /tmp/gcc-arm-none-eabi-linux.tar.bz2 'https://developer.arm.com/-/media/Files/downloads/gnu-rm/7-2018q2/gcc-arm-none-eabi-7-2018-q2-update-linux.tar.bz2?revision=bc2c96c0-14b5-4bb4-9f18-bceb4050fee7?product=GNU%20Arm%20Embedded%20Toolchain%20Downloads,64-bit,,Linux,7-2018-q2-update' &&\
```
Would it be a problem to just use `gcc-arm-none-eabi` and its companion packages directly from Ubuntu repositories?

tamasban commented:
@utzig Do you need any update here?

utzig commented on ci/fih_test_docker/Dockerfile:
```diff
@@ -0,0 +1,58 @@
+# Clone TF-M and dependencies
+RUN mkdir -p /root/work/tfm &&\
+    cd /root/work/tfm  &&\
+    git clone https://git.trustedfirmware.org/TF-M/trusted-firmware-m.git &&\
```
Why do we need this?

utzig commented on ci/fih_test_docker/execute_test.sh:
```diff
@@ -0,0 +1,65 @@
+make -j install
+
+cp ./install/outputs/MPS2/AN521/tfm_s_ns_signed.bin .
+cp ./install/outputs/MPS2/AN521/bl2.axf .
```
What are those files?

utzig reviewed (changes requested):
I'll leave this in progress review here, have not checked all files yet. I am not sure why there is a need for using docker here, since running qemu + gdb on the CI should be entirely fine. Also do we need any trusted-firmware-m dependencies to be able to check for fault injections?

matetothpal commented on ci/fih_test_docker/generate_test_report.py:
```diff
@@ -0,0 +1,87 @@
+
+import argparse
+import yaml
+import argparse
```
I'll fix this

matetothpal commented:
My original idea was that it would be nice to run the tests on an Armv8-M device. However I was not able to make Zephyr work on any of the Qemu supported v8-M platforms, so I decided to use TF-M instead, which has a good support for AN521. TF-M has quite strict requirements on the build environment, and that's why I decided to add a docker layer to the tests, instead running the tests on the CI VMs directly, so I have more direct control.

However I agree that using TF-M might be a bit overkill here. MCUBoot isn't really using any feature of the v8-M architecture, so a v7-M device would do as well. 

@utzig Can you please recommend a v7-M target that is emulated by Qemu and can be used for this FIH testing? Would you recommend using Mynewt or Zephyr?

Once I migrated the scripts to work with another OS and without docker, I'll fix the other issues you flagged as well.

tamasban commented:
Building MCUboot on top of Zephyr or Mynewt is not requiring a docker image? What sample application can be used by MCUboot to verify and boot? It is needed to determine the success of the test. Test script checks whether it  boots up or not.

matetothpal reviewed (commented):
**On using docker:**
As far as I can tell, the Travis CI VM-s are running Ubuntu 16.04. I just tried to run the tests on the VM directly, but it seems that installing the dependencies is not the straightforward. (3.5 python, Quemu not supporting AN521, so python should either be added from a third party package repo, or built from source. Qemu would need to be built from source). Using docker, and installing the necessary SW from the package repo is more straightforward.
**On using TF-M:**
I did some experimenting building Zephyr for a Qemu target, but Ninja seem to have frozen during the build step. Is Do you think Zephyr the way to go anyway?

matetothpal commented:
@utzig I updated the commit with your suggestions. I also added some explanation on the chouces of using Docker, and building with TF-M.

Also please note that as it is now the tests doesn't do much useful, because when building release, the MCUBoot emits no log messages on which the evaluation logic is based on. I'll address this in an update later.

matetothpal commented:
@utzig The latest version fixes the logging issue mentioned in the previous commit

utzig commented on ci/fih_test_docker/Dockerfile:
```diff
@@ -33,21 +36,12 @@ RUN \
-    tar -xf /tmp/gcc-arm-none-eabi-linux.tar.bz2 -C /root/bin/gcc-7-2018-q2-update --strip-components=1 &&\
-    # set paths
-    echo > ~/.bashrc &&\
-    echo 'export PATH=/root/bin/gcc-7-2018-q2-update/bin:$PATH' >> ~/.bashrc
```
The lines that are being removed here were added in the previous commit, so instead fix the previous commit so it does not need to be fixed in a follow up commit.

matetothpal replied:
Done it in the latest push

matetothpal commented:
@utzig Another advantage of using docker is that it is possible to run the tests on a local machine without the need to install any additional software (provided it already has docker and git):

Prerequisite:
The mcuboot to be tested needs to be checked out in a directory hierarchy as it is checked out on the Travis CI VM. so in `<username>/mcuboot`. The mcuboot to be tested is copied into the docker image by the docker build process. Docker build was complaining if the context of the docker image creation is the `<username>/mcuboot` directory, so the `ci/fih-tests_install.sh` script copies it to a directory called `context`, and starts the building from there. This directory needs to be deleted before docker generation. 

So, running the test on the host looks like the following (the current working dir should be the `<username>/mcuboot` directory):

```
$ rm -rf ../../context/*
$ ./ci/fih-tests_install.sh
$ FIH_LEVEL=MCUBOOT_FIH_PROFILE_MEDIUM BUILD_TYPE=RELEASE SKIP_SIZE=2 DAMAGE_TYPE=SIGNATURE ./ci/fih-tests_run.sh
```
On the travis CI the environment variables in the last command are set based on the configs provided in the `.travis.yaml`

This starts the tests, however the shell that it is running in is not interactive, it is not possible to examine the results of the test run. To have an interactive shell where the results can be examined, the following can be done:

0. The docker image needs to be built with `ci/fih-tests_install.sh` as described above
1. Start the docker image with the following command: `$ docker run -i -t mcuboot/fih-test`
2. Execute the test with the following command: `$/root/execute_test.sh 8 RELEASE SIGNATURE MEDIUM`. After the test finishes, the shell returns, and it is possible to investigate the results. It is also possible to stop the test with _Ctrl+c_. The parameters to the `execute_test.sh` are `SKIP_SIZE`, `BUILD_TYPE`, `DAMAGE_TYPE`, `FIH_LEVEL` in order.

matetothpal commented:
@utzig I added some documentation of the FIH testing along with the steps to run the tests locally. I also rebased on the latest master.

tamasban commented:
> > @utzig I added some documentation of the FIH testing along with the steps to run the tests locally. I also rebased on the latest master.
> 
> Thanks, I'll give it a try!

@utzig Have you had chance to take a look and run it on local machine? 
@d3zd3z  Is it still be intended to be merged for v1.7.0 release?

utzig commented:
> @utzig Have you had chance to take a look and run it on local machine?

No.

> @d3zd3z Is it still be intended to be merged for v1.7.0 release?

Yes.

utzig commented on ci/fih_test_docker/damage_image.py:
```diff
@@ -0,0 +1,202 @@
+
+        tlv_off += len(tlv)
+
+def main():
```
The code looks fine, but you should consider installing and running `flake8`. Spacing between root level classes, functions, etc, should be 2 empty lines, you are using 1 almost everywhere here.

utzig reviewed (changes requested):
The build of the docker image was OK, although I had to make a few local changes but left them on the _install.sh script. While trying to run I get:

```
$ docker run mcuboot/fih-test /bin/sh -c '/root/execute_test.sh $0 $1 $2' 2,4,6,8,10 RELEASE SIGNATURE
+ set -e
+ WORKING_DIRECTORY=/root/work/tfm
+ MCUBOOT_PATH=/root/work/tfm/mcuboot
+ TFM_DIR=/root/work/tfm/trusted-firmware-m
+ TFM_BUILD_DIR=/root/work/tfm/trusted-firmware-m/build
+ SKIP_SIZE=2,4,6,8,10
+ BUILD_TYPE=RELEASE
+ DAMAGE_TYPE=SIGNATURE
+ FIH_LEVEL=
+ source /root/.bashrc
++ '[' -z '' ']'
++ return
+ test -z ''
+ CMAKE_FIH_LEVEL=
+ mkdir -p /root/work/tfm/trusted-firmware-m/build
+ cd /root/work/tfm/trusted-firmware-m/build
+ echo 'set(MCUBOOT_LOG_LEVEL "INFO" CACHE STRING "Level of logging to use for MCUboot [OFF, ERROR, WARNING, INFO, DEBUG]" FORCE)'
+ cmake -S .. -B . -DCMAKE_BUILD_TYPE=RELEASE -DCMAKE_TOOLCHAIN_FILE=toolchain_GNUARM.cmake -DTFM_PLATFORM=mps2/an521 -DTEST_NS=ON -DTEST_S=ON -DTFM_PSA_API=ON -DMCUBOOT_PATH=/root/work/tfm/mcuboot -DMCUBOOT_LOG_LEVEL=INFO -DTFM_EXTRA_CONFIG_PATH=extra_config.cmake
CMake Error at config/set_config.cmake:14 (include):
  include could not find load file:

    extra_config.cmake
Call Stack (most recent call first):
  CMakeLists.txt:22 (include)

CMake Deprecation Warning at CMakeLists.txt:41 (message):
  SETTING CMAKE_TOOLCHAIN_FILE is deprecated.  It has been replaced with
  TFM_TOOLCHAIN_FILE.

CMake Deprecation Warning at CMakeLists.txt:42 (message):
  INTERPRETING -DCMAKE_TOOLCHAIN_FILE=toolchain_GNUARM.cmake as
  -DTFM_TOOLCHAIN_FILE=toolchain_GNUARM.cmake

CMake Error at CMakeLists.txt:48 (message):
  SETTING CMAKE_TOOLCHAIN_FILE no longer accepts relative paths.  Please
  supply an absolute path or instead use TFM_TOOLCHAIN_FILE (which does
  accept relative paths)

-- Configuring incomplete, errors occurred!
```

Where does this `extra_config.cmake` file come from?

Maybe you should consider breaking these commits into smaller commits so that it can be reviewed in parts and there is no need to have to be rewriting all of them on every review?

utzig commented:
> Where does this `extra_config.cmake` file come from?

Just started looking at `execute_test.sh` and I know where it came from, hehe, have to check why it failed...

utzig commented:
So I just run this again on Travis, to confirm the issue I had locally, and it was reproduced, see: https://travis-ci.org/github/mcu-tools/mcuboot/builds/739240313

I suspect this must be related to some cmake changes in the TFM repo, which makes quite worried about future maintenance of this feature.  We should probably pick a tag or something that does not move like download a release, etc. I was able to build and run the injections locally by doing the following changes, I suggest you pick at least some of them:

```diff
diff --git a/ci/fih-tests_install.sh b/ci/fih-tests_install.sh
index 6ee6a84a4..bd57a250f 100755
--- a/ci/fih-tests_install.sh
+++ b/ci/fih-tests_install.sh
@@ -16,13 +16,19 @@
 
 set -e
 
-# copy mcuboot so that it is part of the docker build context
-REPO_DIR=`pwd`
-mkdir -p ../../context
-pushd ../../context
-cp -r $REPO_DIR .
+# get mcuboot root; assumes running script is stored under REPO_DIR/ci/
+REPO_DIR=$(dirname $(dirname $(realpath $0)))
+
+pushd $(mktemp -d)
+
+for dir in boot/bootutil ci scripts; do
+  mkdir -p mcuboot/$dir
+  cp -r $REPO_DIR/$dir/* mcuboot/$dir
+done
+
 cp -r $REPO_DIR/ci/fih_test_docker/execute_test.sh .
-cp -r $REPO_DIR/ci/fih_test_docker/build.sh .
 cp -r $REPO_DIR/ci/fih_test_docker/Dockerfile .
-./build.sh
-popd
\ No newline at end of file
+
+./mcuboot/ci/fih_test_docker/build.sh
+
+popd
diff --git a/ci/fih_test_docker/execute_test.sh b/ci/fih_test_docker/execute_test.sh
index 2a495e7ec..452ff50f3 100755
--- a/ci/fih_test_docker/execute_test.sh
+++ b/ci/fih_test_docker/execute_test.sh
@@ -38,11 +38,11 @@ fi
 
 # build TF-M with MCUBoot
 mkdir -p $TFM_BUILD_DIR
-cd $TFM_BUILD_DIR
+cd $TFM_DIR
 echo 'set(MCUBOOT_LOG_LEVEL "INFO" CACHE STRING "Level of logging to use for MCUboot [OFF, ERROR, WARNING, INFO, DEBUG]" FORCE)' > extra_config.cmake
-cmake -S .. -B . \
+cmake -B $TFM_BUILD_DIR \
     -DCMAKE_BUILD_TYPE=$BUILD_TYPE \
-    -DCMAKE_TOOLCHAIN_FILE=toolchain_GNUARM.cmake \
+    -DTFM_TOOLCHAIN_FILE=toolchain_GNUARM.cmake \
     -DTFM_PLATFORM=mps2/an521 \
     -DTEST_NS=ON \
     -DTEST_S=ON \
@@ -50,7 +50,9 @@ cmake -S .. -B . \
     -DMCUBOOT_PATH=$MCUBOOT_PATH \
     -DMCUBOOT_LOG_LEVEL=INFO \
     -DTFM_EXTRA_CONFIG_PATH=extra_config.cmake \
-    $CMAKE_FIH_LEVEL
+    $CMAKE_FIH_LEVEL \
+    .
+cd $TFM_BUILD_DIR
 make -j install
 
 BOOTLOADER_AXF='./install/outputs/MPS2/AN521/bl2.axf'
```

matetothpal commented:
@utzig Thanks for looking at the changes. I'll fix issues you pointed at in the code later this week. 

Your concerns regarding the TF-M version seems to be valid. For now I'll fix the test on the TF-Mv1.2-RC1 tag from last week. We can update then this to the actual release later when available.

Regarding splitting up the change in more commits: I also think smaller commits easier to digest, however I don't know how this could be done with this. I can for example split up the change to 1 commit /file, but that wouldn't help much I think, because the scripts only make sense if are used together, and if I change something, I often need to touch all the scripts. However if you think it would help, I'm happy to do that.

Thanks for the patch, I'll definitely use it.

utzig commented:
> Regarding splitting up the change in more commits: I also think smaller commits easier to digest, however I don't know how this could be done with this. I can for example split up the change to 1 commit /file, but that wouldn't help much I think, because the scripts only make sense if are used together, and if I change something, I often need to touch all the scripts. However if you think it would help, I'm happy to do that.

Documentation could be one commit, Python scripts another, shell scripts and Dockerfile another commit, and changes to .travis.yml in another.

matetothpal commented on docs/design.md:
```diff
@@ -1129,3 +1129,137 @@ this, the target must provide a definition for the `boot_save_shared_data()`
+image was booted by the bootloader or not.
+
+The main idea is that instruction skipping can be automated by scripting a
+debugger to automatically. The steps executed by the script are:
```
will fix

matetothpal commented on docs/design.md:
```diff
@@ -1129,3 +1129,137 @@ this, the target must provide a definition for the `boot_save_shared_data()`
+- continue execution
+- on breakpoint hit increase the Program Counter
+- continue execution.
+- detach from target after a timeout reached.
```
OK

matetothpal commented on docs/design.md:
```diff
@@ -1129,3 +1129,137 @@ this, the target must provide a definition for the `boot_save_shared_data()`
+
+As MCUboot is deployed on a microcontroller, testing FI would not make much
+sense in the simulator environment running on a host machine with different
+architecture than the MCU's, as the degree of hardening much depends on the
```
will fix

matetothpal commented on docs/design.md:
```diff
@@ -1129,3 +1129,137 @@ this, the target must provide a definition for the `boot_save_shared_data()`
+sense in the simulator environment running on a host machine with different
+architecture than the MCU's, as the degree of hardening much depends on the
+compiler behaviour. For example, (a bit counterintuitively) the code produced by
+gcc with `O0` optimisation is more resilient against FI attacks than the code
```
ok

matetothpal commented on docs/design.md:
```diff
@@ -1129,3 +1129,137 @@ this, the target must provide a definition for the `boot_save_shared_data()`
+
+The tests run in a docker container inside the CI VMs, to make it more easy to
+deploy build and test environment (QEMU, compilers, interpreters). The CI VMs
+seems to be using quite old Ubuntu (16.04).
```
I didn't know that

matetothpal commented on docs/design.md:
```diff
@@ -1129,3 +1129,137 @@ this, the target must provide a definition for the `boot_save_shared_data()`
+  # the values provided in the .travis.yaml
+  start_docker_image(skip_sizes, build_type, damage_type, fih_level)
+
+start_docker_image(skip_sizes, build_type, damage_type, fih_level)
```
ok

matetothpal commented on docs/design.md:
````diff
@@ -1129,3 +1129,137 @@ this, the target must provide a definition for the `boot_save_shared_data()`
+  evaluate_run(qemu_log_file)
+```
+
+Further notes:
````
ok

matetothpal commented on docs/design.md:
```diff
@@ -1129,3 +1129,137 @@ this, the target must provide a definition for the `boot_save_shared_data()`
+
+Further notes:
+- The image is corrupted by changing its signature.
+- MCUBOOT_FIH_PROFILE_MAX is not tested as it requires RNG, and the AN521
```
ok

matetothpal commented on docs/design.md:
```diff
@@ -1129,3 +1129,137 @@ this, the target must provide a definition for the `boot_save_shared_data()`
+execution delay to the code, so should not affect the instruction skip results
+too much, because break point is placed at exact address. But in practice this
+makes harder the accurate timing of the attack.
+- The test cases defined in .travis.yml always return passed, if they were
```
ok

matetothpal commented on docs/design.md:
```diff
@@ -1129,3 +1129,137 @@ this, the target must provide a definition for the `boot_save_shared_data()`
+possible to run the tests on a local machine that has git and docker, without
+installing any additional software.
+
+Prerequisite:
```
ok

matetothpal commented on docs/design.md:
```diff
@@ -1129,3 +1129,137 @@ this, the target must provide a definition for the `boot_save_shared_data()`
+build was complaining if the context of the docker image creation is the
+`<username>/mcuboot` directory, so the `ci/fih-tests_install.sh` script copies it
+to a directory called context, and starts the building from there. This directory
+needs to be deleted before docker generation.
```
The prerequisites are not needed. will update accordingly

matetothpal commented:
@utzig I updated the change based on your comments. 

- I'd like to keep the first commit, adding the docker image, the install/run scripts, and a simple execute sh that compiles and runs QEMU with TF-M.
- Then add the scripts that execute the test itself, on an intact payload. At this stage the test outputs the yaml file with the raw results
- Update the travis yaml
- Then add the python script that damages the signature, and use it in the test scripts.
- Then Add the python script that creates the test summary.
- And then add the documentation.

This way all the commits execute something valid in the CI.

utzig commented:
> @utzig I updated the change based on your comments.

It looks good now, once #869 is merged, you'll have to do some clean up of `.travis.yml` and I'll approve afterwards.

utzig commented:
@matetothpal You need to rebase, and you will need to re-add .travis.yml here because I had to remove since it seems to handle no config options as a Ruby project. Just adding it back should be fine.

matetothpal commented:
@utzig I did the rebase. Re-added the .travis.yaml file, but removed the imgtool test as I seem to understand that it has been migrated to GH workflows. 
The tests on travis seem to have started, so I probably haven't messed up the yaml file. Let's see if they pass: https://travis-ci.com/github/matetothpal/mcuboot/builds/199815202

utzig commented on .travis.yml:
```diff
@@ -0,0 +1,87 @@
+    packages:
+      - "python3-pip"
+
+language: rust
```
rust is not required anymore, so this line can be removed.

matetothpal replied:
Will remove unnecessary stuff

utzig commented on .travis.yml:
```diff
@@ -0,0 +1,87 @@
+cache:
+  directories:
+    - $HOME/TOOLCHAIN
+    - cargo
```
This caching is not needed anymore.

matetothpal replied:
Will remove unnecessary stuff

utzig commented on ci/fih_test_docker/fi_tester_gdb.sh:
```diff
@@ -0,0 +1,200 @@
+# echo -ne "$SKIP_ADDRESS | $INSTR...\t"
+
+# Write out a script with the instructions for the debugger to run
+echo "
```
I guess changing this to a here document would make it more readable.

```
cat >commands.gdb <<EOF
target remote localhost: 1234
b *$SKIP_ADDRESS
continue&
eval "shell sleep 0.5"
interrupt
if \$pc == $SKIP_ADDRESS
    echo "Stopped at breakpoint"
else
    echo "Failed to stop at breakpoint"
end
echo "PC before increase:"
print \$pc
set \$pc += $SKIP_SIZE
echo "PC after increase:"
print \$pc
detach
eval "shell sleep 0.5"
EOF
```

matetothpal replied:
Ok

utzig commented on ci/fih_test_docker/fi_tester_gdb.sh:
```diff
@@ -0,0 +1,200 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+function skip_instruction {
```
The body of this function should be indented (apart from the multi-line echo that is...).

matetothpal replied:
Ok. Although the `EOF` added for the next comment cannot be indented, so I left the whole gdb script part unindented.

utzig commented on ci/fih_test_docker/fi_tester_gdb.sh:
```diff
@@ -0,0 +1,200 @@
+# Inform how the script is used
+usage() {
+    echo "$0 <image_dir> <start_addr> [<end_addr>] [(-s | --skip) <skip_len>]"
+} # function skip_instruction
```
Is there a purpose for this comment?

matetothpal replied:
No, I'll remove it

utzig commented on ci/fih_test_docker/fi_tester_gdb.sh:
```diff
@@ -0,0 +1,200 @@
+echo "
+target remote localhost: 1234
+b *$SKIP_ADDRESS
+continue&
```
Is the `&` suffix on purpose?

matetothpal replied:
Yes. It is necessary, in case the breakpoint is not hit (e.g. it is on a conditional branch that is not taken during execution).
The `&` makes the command start in the background, and the `interrupt` command below is executed even if no breakpoint was hit.

matetothpal reviewed (commented):
@utzig Done the changes you requested. 
https://travis-ci.com/github/matetothpal/mcuboot/builds/199908001
